x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/ScopedRole.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.core.security.authz.permission;
+
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.transport.TransportRequest;
+import org.elasticsearch.xpack.core.security.authz.accesscontrol.IndicesAccessControl;
+
+import java.util.Set;
+import java.util.function.Predicate;
+
+/**
+ * A {@link Role} scoped by another role.<br>
+ * The effective permissions returned on {@link #authorize(String, Set, MetaData, FieldPermissionsCache)} call would be scoped by the
+ * provided role.
+ */
+public final class ScopedRole extends Role {
+    private final Role scopedBy;
+
+    ScopedRole(String[] names, ClusterPermission cluster, IndicesPermission indices, ApplicationPermission application,
+            RunAsPermission runAs, Role scopedBy) {
+        super(names, cluster, indices, application, runAs);
+        this.scopedBy = scopedBy;
+    }
+
+    public Role scopedBy() {
+        return scopedBy;
+    }
+
+    @Override
+    public IndicesAccessControl authorize(String action, Set<String> requestedIndicesOrAliases, MetaData metaData,
+            FieldPermissionsCache fieldPermissionsCache) {
+        IndicesAccessControl indicesAccessControl = super.authorize(action, requestedIndicesOrAliases, metaData, fieldPermissionsCache);
+        IndicesAccessControl scopedByIndicesAccessControl = scopedBy.authorize(action, requestedIndicesOrAliases, metaData,
+                fieldPermissionsCache);
+
+        return IndicesAccessControl.scopedIndicesAccessControl(indicesAccessControl, scopedByIndicesAccessControl);
+    }
+
+    /**
+     * @return A predicate that will match all the indices that this role and
+     * the scoped by role has the privilege for executing the given action on.
+     */
+    public Predicate<String> allowedIndicesMatcher(String action) {
+        Predicate<String> predicate = indices().allowedIndicesMatcher(action);
+        if (scopedBy != null) {
```

Comment:
Wouldn't it be better to simply require that a `ScopedRole` must have a `scopedBy` ? Otherwise it's just a `Role`.

Reply:
👍